src/cpu/o3/comm.cc

```diff
@@ -0,0 +1,74 @@
+#include "cpu/o3/comm.hh"
+
+#include <algorithm>
+
+#include "cpu/o3/dyn_inst.hh"
+
+namespace gem5
+{
+
+namespace o3
+{
+
+/**
+ * Remove instructions belonging to given thread from plain
+ * instruction array. Automatically updates the array size.
+ */
+static void
+removeThreadInsts(ThreadID tid, DynInstPtr *insts, int &size)
+{
+    auto has_tid = [tid] (const DynInstPtr &inst) -> bool {
+        return inst->threadNumber == tid;
+    };
+    DynInstPtr *last = std::remove_if(insts, insts + size, has_tid);
```

Comment:
How is this working? insts is actually pointing to a C-style array of DynInstPtr. What will remove_if actually do? I though it was supposed to be used with mutable types like std::vector. I have the gut feeling this might be wrong

Reply:
Sorry for the slow response. Correct me if I'm wrong, but I think you might be thinking of [std::erase_if](https://en.cppreference.com/w/cpp/container/vector/erase2), which operates on C++ containers like std::vector and std::map? My understanding is that [std::remove_if](https://en.cppreference.com/w/cpp/algorithm/remove) works on arbitrary ranges, including C-style arrays. How std::remove_if works is for each element to be removed, it moves (specifically using a move assignment) an element from the end of the array to the index of the remove element.

Reply:
You are right. However I was wondering whether pushing at the end of the array is enough. Shouldn't we de-allocate the dynamic instruction as well to avoid memory leaks?